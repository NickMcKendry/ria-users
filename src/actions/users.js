import database from '../firebase/firebase';

// ADD_USER

export const addUser = (user) => ({
  type: 'ADD_USER',
  user
});

export const startAddUser = (userData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      firstName = '',
      lastName = '',
      address = 0,
      createdAt = 0
    } = userData;

    const user = {
      firstName,
      lastName,
      address,
      createdAt
    };

    return database.ref(`users/${uid}/users`).push(user).then((ref) => {
      dispatch(addUser({
        id: ref.key,
        ...user
      }));
    });
  };
};

// REMOVE_USER

export const removeUser = ({ id } = {}) => ({
  type: 'REMOVE_USER',
  id
});

export const startRemoveUser = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users/${id}`).remove().then(() => {
      dispatch(removeUser({
        id
      }))
    })
  }
};

// EDIT_USER

export const editUser = (id, updates) => ({
  type: 'EDIT_USER',
  id,
  updates
});

export const startEditUser = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users/${id}`).update(updates).then(() => {
      dispatch(editUser(id, updates))
    });
  };
};

// SET_USERS

export const setUsers = (users) => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users`).once('value').then((snapshot) => {
      const users = [];

      snapshot.forEach((childSnapshot) => {
        users.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setUsers(users));
    });
  };
};
