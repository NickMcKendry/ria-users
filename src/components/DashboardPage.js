import React from 'react';
import ReactDOM from 'react-dom';

const DashboardPage = () => (
  <div>
    <p>DashboardPage content</p>
  </div>
);

export default DashboardPage;
