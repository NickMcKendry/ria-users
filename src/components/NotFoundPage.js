import React from 'react';
import ReactDOM from 'react-dom';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
  <div>
    <h3>404 - <Link to="/">Go Home</Link></h3>
  </div>
);

export default NotFoundPage;
