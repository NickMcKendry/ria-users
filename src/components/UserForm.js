import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class UserForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      firstName: props.user ? props.user.firstName : '',
      lastName: props.user ? props.user.lastName : '',
      address: props.user ? props.expense.address : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  };


  onFirstNameChange = (e) => {
    const firstName = e.target.value;
    this.setState(() => ({ firstName }));
  };

  onLastNameChange = (e) => {
    const lastName = e.target.value;
    this.setState(() => ({ lastName }));
  };

  onAddressChange = (e) => {
    const address = e.target.value;
    this.setState(() => ({ address }))
  };

  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  };

  onSubmit = ((e) => {
    e.preventDefault();

    if (!this.state.firstName || !this.state.lastName || !this.state.address) {
      this.setState(() => ({ error: 'Please provide a firstName, lastName and address!' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        firstName: this.state.firstName,
        lastName: this.state.lastName,
        createdAt: this.state.createdAt.valueOf(),
        address: this.state.address
      })
    }
  })

  render() {
    return (
        <form onSubmit={this.onSubmit} className="form">
          { this.state.error && <p className="form__error">{this.state.error}</p> }
          <input
            type="text"
            placeholder="First Name"
            autoFocus
            value={this.state.firstName}
            onChange={this.onFirstNameChange}
            className="text-input"
          />
          <input
            type="text"
            placeholder="Last Name"
            value={this.state.lastName}
            onChange={this.onLastNameChange}
            className="text-input"
          />
          <input
            type="text"
            placeholder="Address"
            value={this.state.address}
            onChange={this.onAddressChange}
            className="text-input"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
            <div>
              <button className="button">Save User</button>
            </div>

        </form>
    )
  }
};
