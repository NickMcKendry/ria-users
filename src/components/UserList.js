import React from 'react';
import { connect } from 'react-redux';
import UserListItem from './UserListItem';
import selectUsers from '../selectors/users';

export const UserList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Users</div>
      <div className="show-for-desktop">User</div>
      <div className="show-for-desktop">Address</div>
    </div>
    <div className="list-body">
    {
      props.users.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No users</span>
        </div>
      ) : (
        props.users.map((user) => (
          <UserListItem key={user.id} {...user} />
        ))
      )
    }
  </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    users: selectUsers(state.users, state.filters)
  }
};

export default connect(mapStateToProps)(UserList);
