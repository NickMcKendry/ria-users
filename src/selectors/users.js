import moment from 'moment';

export default (users, { text, sortBy, birthDate, endDate }) => {

  return users.filter((user) => {
    const createdAtMoment = moment(user.createdAt)

    const birthDateMatch = birthDate ? birthDate.isSameOrBefore(createdAtMoment, 'day') : true;

    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    const name = user.firstName + user.lastName

    const textMatch = name.toLowerCase().includes(text.toLowerCase());

    return birthDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if(sortBy === 'first-name') {
      return a.amount < b.amount ? 1 : -1;
    } else if(sortBy === 'last-name') {
      return a.amount < b.amount ? 1 : -1;
    } else if(sortBy === 'address') {
      return a.amount < b.amount ? 1 : -1;
    }
  })
}
