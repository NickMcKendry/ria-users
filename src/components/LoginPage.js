import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Boilerplate v2</h1>
      <p>App Tagline</p>
      <button onClick={startLogin} className="button">Login With Google</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
