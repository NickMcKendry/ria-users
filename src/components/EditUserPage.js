import React from 'react';
import UserForm from './UserForm';
import { connect } from 'react-redux';
import { startEditUser, startRemoveUser } from '../actions/users';

export class EditUserPage extends React.Component {


  onSubmit = (user) => {
    this.props.editUser(this.props.user.id, user);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.removeUser({ id: this.props.user.id });
    this.props.history.push('/');
  }

  render() {
    return(
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit user</h1>
          </div>
        </div>
        <div className="content-container">
          <UserForm
            user={this.props.user}
            onSubmit={this.onSubmit}
          />
          <button onClick={this.onClick} className="button button--secondary">Remove User</button>
        </div>
      </div>
    );
  };
};

const mapStateToProps = (state, props) => {
  return {
    user: state.users.find((user) => user.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editUser: (id, user) => dispatch(startEditUser(id, user)),
  removeUser: (id) => dispatch(startRemoveUser(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditUserPage);
