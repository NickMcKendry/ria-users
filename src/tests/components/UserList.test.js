import React from 'react';
import { UserList } from '../../components/UserList';
import { shallow } from 'enzyme';
import users from '../fixtures/users';

test('Should render user list with users', () => {
  const wrapper = shallow(<UserList users={users} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render user list with no users', () => {
  const wrapper = shallow(<UserList users={[]} />);
  expect(wrapper).toMatchSnapshot();
});
