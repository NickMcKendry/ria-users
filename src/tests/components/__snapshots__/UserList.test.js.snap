// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render user list with no users 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Users
    </div>
    <div
      className="show-for-desktop"
    >
      User
    </div>
    <div
      className="show-for-desktop"
    >
      Address
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No users
      </span>
    </div>
  </div>
</div>
`;

exports[`Should render user list with users 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Users
    </div>
    <div
      className="show-for-desktop"
    >
      User
    </div>
    <div
      className="show-for-desktop"
    >
      Address
    </div>
  </div>
  <div
    className="list-body"
  >
    <UserListItem
      address="9803 w. 83rd ave."
      firstName="Nick"
      id="1"
      key="1"
      lastName="McKendry"
    />
    <UserListItem
      address="9803 w. 83rd ave."
      firstName="John"
      id="2"
      key="2"
      lastName="Doe"
    />
    <UserListItem
      address="9803 w. 83rd ave."
      firstName="Jane"
      id="3"
      key="3"
      lastName="Dae"
    />
  </div>
</div>
`;
